File: apps/ios/Shared/Views/Badges/BadgesPayView.swift
//
//  BadgesPayView.swift
//  SimpleX (iOS)
//
//  Created by spaced4ndy on 28.07.2026.
//  Copyright © 2026 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat

// TODO [badges]: replace with types produced by the badge purchase API when it lands.
enum BadgePeriod: String, CaseIterable, Identifiable {
    case oneMonth
    case monthly
    case annual

    var id: String { rawValue }

    var icon: String {
        switch self {
        case .oneMonth: "calendar"
        case .monthly: "arrow.clockwise"
        case .annual: "arrow.clockwise"
        }
    }

    var label: LocalizedStringKey {
        switch self {
        case .oneMonth: "1 month"
        case .monthly: "Monthly"
        case .annual: "Annual"
        }
    }

    func priceText(_ price: BadgePrice) -> Text {
        switch price {
        case .loading: return Text(verbatim: "…")
        case .unavailable: return Text(verbatim: "—")
        case let .price(p):
            switch self {
            case .oneMonth: return Text(verbatim: p)
            case .monthly: return Text("\(p)/month")
            case .annual: return Text("\(p)/year")
            }
        }
    }

    func payText(_ price: BadgePrice) -> Text {
        switch price {
        case .loading: return Text("Loading…")
        case .unavailable: return Text("Not available")
        case let .price(p):
            switch self {
            case .oneMonth: return Text("Pay \(p)")
            case .monthly: return Text("Pay \(p)/month")
            case .annual: return Text("Pay \(p)/year")
            }
        }
    }
}

struct BadgesPayView: View {
    @EnvironmentObject var theme: AppTheme
    @ObservedObject private var store = BadgeStore.shared
    let level: BadgeLevel
    @State private var selectedPeriod: BadgePeriod = .monthly
    @State private var purchasing = false
    // presented from this view, not AlertManager: its host is behind the sheet these views open in
    @State private var alert: SomeAlert?

    var body: some View {
        GeometryReader { g in
            ScrollView {
                VStack(alignment: .center, spacing: 16) {
                    Text(level.title)
                        .font(.largeTitle)
                        .bold()
                        .foregroundColor(theme.colors.primary)
                        .multilineTextAlignment(.center)
                        .fixedSize(horizontal: false, vertical: true)

                    BadgeUserPreview(level: level)
                        .padding(.top, 4)

                    Text(level.tagline)
                        .font(.body)
                        .foregroundColor(theme.colors.onBackground)
                        .multilineTextAlignment(.center)
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.top, 4)

                    Spacer(minLength: 20)

                    // fixedSize + maxHeight on the cards so all three match the tallest one -
                    // only Annual carries a savings line, and prices wrap at large fonts
                    HStack(alignment: .top, spacing: 12) {
                        periodCard(.oneMonth)
                        periodCard(.monthly)
                        periodCard(.annual)
                    }
                    .fixedSize(horizontal: false, vertical: true)

                    Spacer(minLength: 20)

                    VStack(spacing: 10) {
                        payButton()
                            .padding(.vertical, 10)
                        Text(billingFooter)
                            .font(.footnote)
                            .foregroundColor(theme.colors.secondary)
                            .multilineTextAlignment(.center)
                            .fixedSize(horizontal: false, vertical: true)
                            .frame(height: 22)
                    }
                    .padding(.bottom, g.safeAreaInsets.bottom == 0 ? 20 : 0)
                }
                .padding(.horizontal, 25)
                .padding(.top, 0)
                .padding(.bottom, 20)
                .frame(minHeight: g.size.height)
            }
        }
        .frame(maxHeight: .infinity)
        .navigationBarTitleDisplayMode(.inline)
        .task { await store.load() }
        .alert(item: $alert) { $0.alert }
    }

    private func periodCard(_ period: BadgePeriod) -> some View {
        let isSelected = period == selectedPeriod
        return Button {
            selectedPeriod = period
        } label: {
            VStack(spacing: 12) {
                Image(systemName: period.icon)
                    .resizable()
                    .scaledToFit()
                    .frame(width: 32, height: 32)
                    .foregroundColor(isSelected ? theme.colors.primary : theme.colors.secondary)
                Text(period.label)
                    .font(.title3)
                    .fontWeight(.bold)
                period.priceText(store.price(level, period))
                    .font(.body)
                if let percent = savingsPercent(period) {
                    Text("Save \(percent)%")
                        .font(.footnote)
                        .foregroundColor(isSelected ? theme.colors.primary : theme.colors.secondary)
                }
            }
            .multilineTextAlignment(.center)
            .padding(.vertical, 30)
            .padding(.horizontal, 12)
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
            .background(Color(uiColor: .secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 16))
            .overlay(
                RoundedRectangle(cornerRadius: 16)
                    .stroke(isSelected ? theme.colors.primary : Color(uiColor: .secondarySystemFill), lineWidth: 2)
            )
        }
        .buttonStyle(.plain)
    }

    private func savingsPercent(_ period: BadgePeriod) -> Int? {
        period == .annual ? store.annualSavings(level) : nil
    }

    private func payButton() -> some View {
        let price = store.price(level, selectedPeriod)
        let disabled = !price.canPurchase || purchasing
        return Button {
            purchase()
        } label: {
            selectedPeriod.payText(price)
        }
        .buttonStyle(OnboardingButtonStyle(isDisabled: disabled))
        .disabled(disabled)
    }

    private func purchase() {
        let period = selectedPeriod
        let invoiceId = newBadgeInvoiceId()
        purchasing = true
        Task {
            do {
                let outcome = try await store.purchase(level, period, invoiceId: invoiceId)
                await MainActor.run {
                    purchasing = false
                    switch outcome {
                    case let .purchased(receipt): showPurchasedAlert(receipt, invoiceId)
                    case .pending:
                        alert = SomeAlert(
                            alert: mkAlert(
                                title: "Purchase pending",
                                message: "The purchase is awaiting approval. This build does not deliver purchases approved later."
                            ),
                            id: "badgePurchasePending"
                        )
                    case .cancelled: break
                    }
                }
            } catch let error {
                logger.error("BadgesPayView.purchase: \(String(describing: error))")
                await MainActor.run {
                    purchasing = false
                    alert = SomeAlert(
                        alert: Alert(
                            title: Text("Purchase error"),
                            message: Text(verbatim: String(describing: error))
                        ),
                        id: "badgePurchaseError"
                    )
                }
            }
        }
    }

    // TODO [badges] store integration diagnostics - replaced by the issued badge once the service lands.
    private func showPurchasedAlert(_ receipt: BadgeStoreReceipt, _ invoiceId: UUID) {
        let returnedInvoice: String
        if let returned = receipt.invoiceId {
            returnedInvoice = returned == invoiceId ? "yes" : "mismatch: \(returned.uuidString)"
        } else {
            returnedInvoice = "none"
        }
        var lines = [
            "Product: \(receipt.productId)",
            "Invoice: \(invoiceId.uuidString)",
            "Invoice returned by Apple: \(returnedInvoice)",
            "Transaction: \(receipt.transactionId)"
        ]
        if let environment = receipt.environment { lines.append("Environment: \(environment)") }
        lines.append("Signature: \(receipt.signatureVerified ? "verified" : "unverified")")
        lines.append("Token: \(receipt.jws.count) bytes")
        let summary = lines.joined(separator: "\n")
        // logged as well as shown: the alert races StoreKit's own sheets, the log always lands
        logger.debug("badge purchase succeeded\n\(summary)")
        alert = SomeAlert(
            alert: Alert(
                title: Text("Purchase successful"),
                message: Text(verbatim: summary),
                primaryButton: .default(Text(verbatim: "Copy token")) { UIPasteboard.general.string = receipt.jws },
                secondaryButton: .cancel(Text("Ok"))
            ),
            id: "badgePurchased"
        )
    }

    private var billingFooter: LocalizedStringKey {
        // TODO [badges] source the actual date from the purchase state machine when wired.
        var comps = DateComponents(); comps.year = 2026; comps.month = 7; comps.day = 22
        let stubDate = Calendar.current.date(from: comps) ?? Date()
        let date = DateFormatter.localizedString(from: stubDate, dateStyle: .long, timeStyle: .none)
        switch selectedPeriod {
        case .monthly, .annual: return "Renews on \(date). Cancel anytime."
        case .oneMonth: return "Ends on \(date)."
        }
    }
}

struct BadgesPayView_Previews: PreviewProvider {
    static var previews: some View {
        NavigationView {
            BadgesPayView(level: .supporter)
        }
    }
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/badges/BadgesPayView.kt
package chat.simplex.common.views.badges

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.ClipboardManager
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import dev.icerock.moko.resources.StringResource
import dev.icerock.moko.resources.compose.painterResource
import dev.icerock.moko.resources.compose.stringResource
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.helpers.*
import chat.simplex.common.views.onboarding.OnboardingActionButton
import chat.simplex.res.MR

// TODO [badges]: replace with types produced by the badge purchase API when it lands.
enum class BadgePeriod {
  OneMonth,
  Monthly,
  Annual;

  val icon: dev.icerock.moko.resources.ImageResource
    get() = when (this) {
      OneMonth -> MR.images.ic_calendar
      Monthly -> MR.images.ic_refresh
      Annual -> MR.images.ic_refresh
    }

  val label: StringResource
    get() = when (this) {
      OneMonth -> MR.strings.badges_period_one_month
      Monthly -> MR.strings.badges_period_monthly
      Annual -> MR.strings.badges_period_annual
    }

  @Composable
  fun priceText(price: BadgePrice): String = when (price) {
    is BadgePrice.Loading -> "…"
    is BadgePrice.Unavailable -> "—"
    is BadgePrice.Price -> when (this) {
      OneMonth -> price.price
      Monthly -> stringResource(MR.strings.badges_price_monthly).format(price.price)
      Annual -> stringResource(MR.strings.badges_price_annual).format(price.price)
    }
  }

  // OnboardingActionButton takes a resource id, so the price states map to (id, arg) here rather
  // than to a formatted string as on iOS
  fun payLabel(price: BadgePrice): Pair<StringResource, String?> = when (price) {
    is BadgePrice.Loading -> MR.strings.badges_price_loading to null
    is BadgePrice.Unavailable -> MR.strings.badges_price_unavailable to null
    is BadgePrice.Price -> when (this) {
      OneMonth -> MR.strings.badges_pay_once to price.price
      Monthly -> MR.strings.badges_pay_monthly to price.price
      Annual -> MR.strings.badges_pay_annual to price.price
    }
  }
}

@Composable
fun BadgesPayView(level: BadgeLevel) {
  var selectedPeriod by remember { mutableStateOf(BadgePeriod.Monthly) }
  val purchasing = remember { mutableStateOf(false) }
  val clipboard = LocalClipboardManager.current

  LaunchedEffect(Unit) { BadgeStore.load() }

  ColumnWithScrollBar(
    Modifier.background(MaterialTheme.colors.background).padding(horizontal = 25.dp).padding(top = 8.dp, bottom = 20.dp),
    verticalArrangement = Arrangement.spacedBy(16.dp),
    horizontalAlignment = Alignment.CenterHorizontally,
    maxIntrinsicSize = true,
  ) {
    Text(
      stringResource(level.title),
      style = MaterialTheme.typography.h1,
      fontWeight = FontWeight.Bold,
      color = MaterialTheme.colors.primary,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth()
    )

    BadgeUserPreview(level = level, modifier = Modifier.padding(top = 4.dp))

    Text(
      stringResource(level.tagline),
      style = MaterialTheme.typography.body1,
      color = MaterialTheme.colors.onBackground,
      textAlign = TextAlign.Center,
      modifier = Modifier.fillMaxWidth().padding(top = 4.dp)
    )

    Spacer(Modifier.weight(1f).heightIn(min = 20.dp))

    // IntrinsicSize.Max + fillMaxHeight on children so all three cards match the tallest one -
    // only Annual carries a savings line, and prices wrap at large fonts.
    Row(
      Modifier.fillMaxWidth().height(IntrinsicSize.Max),
      horizontalArrangement = Arrangement.spacedBy(12.dp)
    ) {
      PeriodCard(level, BadgePeriod.OneMonth, selectedPeriod, Modifier.weight(1f).fillMaxHeight()) { selectedPeriod = it }
      PeriodCard(level, BadgePeriod.Monthly, selectedPeriod, Modifier.weight(1f).fillMaxHeight()) { selectedPeriod = it }
      PeriodCard(level, BadgePeriod.Annual, selectedPeriod, Modifier.weight(1f).fillMaxHeight()) { selectedPeriod = it }
    }

    Spacer(Modifier.weight(1f).heightIn(min = 20.dp))

    // Replicates TextButtonBelowOnboardingButton spacing (7.5dp outer + 5dp inner) without a
    // TextButton so the footer has no hover/click affordance.
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
      PayButton(level, selectedPeriod, purchasing, clipboard)
      Box(Modifier.padding(top = 7.5.dp, bottom = 7.5.dp).padding(horizontal = 16.dp, vertical = 8.dp)) {
        Text(
          stringResource(billingFooter(selectedPeriod)).format(stubBillingDate()),
          Modifier.padding(vertical = 5.dp),
          style = MaterialTheme.typography.body2,
          color = MaterialTheme.colors.secondary,
          textAlign = TextAlign.Center
        )
      }
    }
  }
}

@Composable
private fun PeriodCard(level: BadgeLevel, period: BadgePeriod, selectedPeriod: BadgePeriod, modifier: Modifier, onSelect: (BadgePeriod) -> Unit) {
  val isSelected = period == selectedPeriod
  val borderColor = if (isSelected) MaterialTheme.colors.primary else MaterialTheme.colors.background.mixWith(MaterialTheme.colors.onBackground, 0.92f)
  // Light: transparent so card matches page background. Dark: subtle gray tint for visible contrast.
  val cardBackground = if (isInDarkTheme()) MaterialTheme.colors.background.mixWith(MaterialTheme.colors.onBackground, 0.97f)
                       else MaterialTheme.colors.background
  val shape = RoundedCornerShape(16.dp)
  Column(
    modifier
      .clip(shape)
      .background(cardBackground, shape)
      .border(2.dp, borderColor, shape)
      .clickable { onSelect(period) }
      .padding(vertical = 30.dp, horizontal = 12.dp),
    horizontalAlignment = Alignment.CenterHorizontally,
    verticalArrangement = Arrangement.spacedBy(12.dp)
  ) {
    Icon(
      painterResource(period.icon),
      contentDescription = null,
      tint = if (isSelected) MaterialTheme.colors.primary else MaterialTheme.colors.secondary,
      modifier = Modifier.size(32.dp)
    )
    Text(stringResource(period.label), style = MaterialTheme.typography.h3, fontWeight = FontWeight.Bold, textAlign = TextAlign.Center)
    Text(period.priceText(BadgeStore.price(level, period)), style = MaterialTheme.typography.body1, textAlign = TextAlign.Center)
    val percent = savingsPercent(level, period)
    if (percent != null) {
      Text(
        stringResource(MR.strings.badges_savings).format(percent),
        style = MaterialTheme.typography.caption,
        color = if (isSelected) MaterialTheme.colors.primary else MaterialTheme.colors.secondary,
        textAlign = TextAlign.Center
      )
    }
  }
}

private fun savingsPercent(level: BadgeLevel, period: BadgePeriod): Int? =
  if (period == BadgePeriod.Annual) BadgeStore.annualSavings(level) else null

@Composable
private fun PayButton(level: BadgeLevel, selectedPeriod: BadgePeriod, purchasing: MutableState<Boolean>, clipboard: ClipboardManager) {
  val price = BadgeStore.price(level, selectedPeriod)
  val (labelId, labelArg) = selectedPeriod.payLabel(price)
  OnboardingActionButton(
    modifier = if (appPlatform.isAndroid) Modifier.padding(horizontal = DEFAULT_ONBOARDING_HORIZONTAL_PADDING).fillMaxWidth() else Modifier.widthIn(min = 300.dp),
    labelId = labelId,
    labelArg = labelArg,
    onboarding = null,
    enabled = price.canPurchase && !purchasing.value,
    onclick = { purchase(level, selectedPeriod, purchasing, clipboard) }
  )
}

private fun purchase(level: BadgeLevel, period: BadgePeriod, purchasing: MutableState<Boolean>, clipboard: ClipboardManager) {
  val invoiceId = newBadgeInvoiceId()
  purchasing.value = true
  // not withBGApi: the purchase waits for the user in the Play sheet and would block chat API calls
  withLongRunningApi {
    try {
      val outcome = BadgeStore.purchase(level, period, invoiceId)
      purchasing.value = false
      when (outcome) {
        is BadgePurchaseOutcome.Purchased -> showPurchasedAlert(outcome.receipt, invoiceId, clipboard)
        is BadgePurchaseOutcome.Pending -> AlertManager.shared.showAlertMsg(
          title = generalGetString(MR.strings.badges_purchase_pending),
          text = generalGetString(MR.strings.badges_purchase_pending_desc)
        )
        is BadgePurchaseOutcome.Cancelled -> {}
      }
    } catch (e: Exception) {
      Log.e(TAG, "BadgesPayView.purchase: ${e.stackTraceToString()}")
      purchasing.value = false
      AlertManager.shared.showAlertMsg(
        title = generalGetString(MR.strings.badges_purchase_error),
        text = e.toString()
      )
    }
  }
}

// TODO [badges] store integration diagnostics - replaced by the issued badge once the service lands.
private fun showPurchasedAlert(receipt: BadgeStoreReceipt, invoiceId: String, clipboard: ClipboardManager) {
  val returnedInvoice = when (receipt.invoiceId) {
    null -> "none"
    invoiceId -> "yes"
    else -> "mismatch: ${receipt.invoiceId}"
  }
  val summary = listOf(
    "Product: ${receipt.productId}",
    "Invoice: $invoiceId",
    "Invoice returned by Google: $returnedInvoice",
    "Order: ${receipt.orderId ?: "none"}",
    "Token: ${receipt.token.length} bytes"
  ).joinToString("\n")
  // logged as well as shown: the log always lands even when the alert is missed
  Log.d(TAG, "badge purchase succeeded\n$summary")
  AlertManager.shared.showAlertDialog(
    title = generalGetString(MR.strings.badges_purchase_successful),
    text = summary,
    confirmText = "Copy token",
    onConfirm = {
      clipboard.setText(AnnotatedString(receipt.token))
      showToast(generalGetString(MR.strings.copied))
    },
    dismissText = generalGetString(MR.strings.ok),
    parseHtml = false
  )
}

private fun billingFooter(period: BadgePeriod): StringResource = when (period) {
  BadgePeriod.Monthly, BadgePeriod.Annual -> MR.strings.badges_billing_footer_subscribe
  BadgePeriod.OneMonth -> MR.strings.badges_billing_footer_one_month
}

// TODO [badges] source the actual date from the purchase state machine when wired.
private fun stubBillingDate(): String {
  val date = java.time.LocalDate.of(2026, 7, 22)
  val formatter = java.time.format.DateTimeFormatter.ofLocalizedDate(java.time.format.FormatStyle.LONG)
  return date.format(formatter)
}
